### ISSUE TITLE
bcftools plugin fill-tags segfaults if no pipe exists

### ISSUE THREAD

**Issue by ghuls** [samtools/bcftools#1156]
`bcftools plugin fill-tags` segfaults if no pipe exists.

```
# bcftools plugin fill-tags segfaults if it can't read stdin. 
$ bcftools plugin fill-tags --threads 8 --output-type b -o "x.bcf" -- -t all
Segmentation fault (core dumped)
```

```
# It works when there is a pipe.
$ echo | bcftools plugin fill-tags --threads 8 --output-type b -o "x.bcf" -- -t all
Failed to read from standard input: unknown file type

$ printf '' | bcftools +fill-tags --threads 8 --output-type b -o "x.bcf" -- -t all
Failed to read from standard input: unknown file type
```

Also is there a reason that `samtools view` supports `--output-file` while most (all?) other tools have `--output` instead? Wouldn't it be more consistent that they support `--output-file` too?

**Comment by jmarshall** [comment]
Re `bcftools view --output-file`: the [convention being followed](https://www.gnu.org/prep/standards/html_node/Option-Table.html) is `--output`, but unfortunately view also has `--output-type` and

```
$ bcftools view --output x.bcf
bcftools: option `--output' is ambiguous
```

**Comment by ghuls** [comment]
But all other bcftools subcommand also have `--output` and `--output-type`:

```
$ bcftools call 2>&1 | grep -- --output
   -o, --output <file>             write output to a file [standard output]
   -O, --output-type <b|u|z|v>     output type: 'b' compressed BCF; 'u' uncompressed BCF; 'z' compressed VCF; 'v' uncompressed VCF [v]
```

**Comment by jmarshall** [comment]
Good point. So I guess this `--output-file` is a historical hypercorrection to over-avoid this ambiguity, and the good news is `view` could be brought into line with the others via

```diff
@@ -501,7 +501,7 @@ static void usage(args_t *args)
     fprintf(stderr, "    -h/H, --header-only/--no-header     print the header only/suppress the header in VCF output\n");
     fprintf(stderr, "    -l,   --compression-level [0-9]     compression level: 0 uncompressed, 1 best speed, 9 best compression [%d]\n", args->clevel);
     fprintf(stderr, "          --no-version                  do not append version and command line to the header\n");
-    fprintf(stderr, "    -o,   --output-file <file>          output file name [stdout]\n");
+    fprintf(stderr, "    -o,   --output <file>               output file name [stdout]\n");
     fprintf(stderr, "    -O,   --output-type <b|u|z|v>       b: compressed BCF, u: uncompressed BCF, z: compressed VCF, v: uncompressed VCF [v]\n");
     fprintf(stderr, "    -r, --regions <region>              restrict to comma-separated list of regions\n");
     fprintf(stderr, "    -R, --regions-file <file>           restrict to regions listed in a file\n");
@@ -574,6 +574,7 @@ int main_vcfview(int argc, char *argv[])
         {"samples-file",required_argument,NULL,'S'},
         {"force-samples",no_argument,NULL,1},
         {"output-type",required_argument,NULL,'O'},
+        {"output",required_argument,NULL,'o'},
         {"output-file",required_argument,NULL,'o'},
         {"types",required_argument,NULL,'v'},
         {"exclude-types",required_argument,NULL,'V'},
```

Hopefully the other half of this issue is also so easily dealt with :smile:

**Comment by ghuls** [comment]
`samtools sort` seems to behave the same as `samtools view`.

**Comment by pd3** [comment]
Thanks for the bug report, the segfault should be fixed by 3f96f33a47e.  Also the option `--output-file` was renamed to `--output` to make it consistent across bcftools commands.
